File: pippip-native-ios/API/ServerAuthorized.m
//
//  ServerAuthorized.m
//  pippip-native-ios
//
//  Created by Steve Brenneis on 12/6/17.
//  Copyright © 2017 seComm. All rights reserved.
//

#import "pippip_native_ios-Swift.h"
#import "ServerAuthorized.h"
#import "ApplicationSingleton.h"
#import "AccountManager.h"
#import "CKRSACodec.h"

@interface ServerAuthorized ()
{
    
    SessionState *sessionState;
    
}

@end

@implementation ServerAuthorized

- (instancetype)init {
    self = [super init];
    
    sessionState = [[SessionState alloc] init];
    return self;
    
}

- (NSDictionary*)restPacket {
    
    NSMutableDictionary *packet = [[NSMutableDictionary alloc] init];

    packet[@"sessionId"] = [NSNumber numberWithInt:sessionState.sessionId];
    CKRSACodec *codec = [[CKRSACodec alloc] init];
    [codec putBlock:sessionState.enclaveKey];
    NSData *data = [codec encrypt:sessionState.serverPublicKey];
    packet[@"data"] = [data base64EncodedStringWithOptions:0];
#if !TARGET_OS_SIMULATOR
    AccountSession *accountSession = [ApplicationSingleton instance].accountSession;
    packet[@"deviceToken"] = [accountSession.deviceToken base64EncodedStringWithOptions:0];
#endif
#ifdef DEBUG
    packet[@"developer"] = [NSNumber numberWithBool:YES];
#else
    packet[@"developer"] = [NSNumber numberWithBool:NO];
#endif

    return packet;
    
}

- (double)restTimeout {
    return 15.0;
}

- (NSString*)restPath {
    
    if (AccountManager.production) {
        return @"/authenticator/authorized";
    }
    else {
        return @"/authorized";
    }

}

@end

File: pippip-native-ios/API/ServerAuthorized.swift
//
//  ServerAuthorized.swift
//  pippip-native-ios
//
//  Created by Steve Brenneis on 5/29/18.
//  Copyright © 2018 seComm. All rights reserved.
//

import UIKit
import ObjectMapper

class ServerAuthorized: NSObject, APIRequestProtocol {
    
    var path: String {
        if AccountSession.production {
            return "/authenticator/authorized";
        }
        else {
            return "/authorized";
        }
    }
    var timeout: Double = 10.0
    var sessionId: Int32?
    var authToken: Int64?
    var data: String?
    var deviceToken: String?
    var developer: Bool?
    
    var sessionState = SessionState()

    override init() {

        sessionId = sessionState.sessionId
        let codec = CKRSACodec()
        codec.putBlock(sessionState.enclaveKey!)
        let encoded = codec.encrypt(sessionState.serverPublicKey!)
        data = encoded.base64EncodedString()
        // Will be nil if simulator
        deviceToken = AccountSession.instance.deviceToken?.base64EncodedString()
        #if DEBUG
        developer = true
        #else
        developer = false
        #endif

        super.init()

    }

    required init?(map: Map) {
        
    }
    
    func mapping(map: Map) {

        sessionId <- map["sessionId"]
        data <- map["data"]
        deviceToken <- map["deviceToken"]
        developer <- map["developer"]

    }
    
}
